cip/CIP2016-06-14-Harmonize-comparability-orderability-equality-and-equivalence.adoc

```diff
@@ -0,0 +1,233 @@
+
+Any additionally introduced non-canonical types introduced by an implementation should be inserted between `PATH` and `MAP` in the global sort order. As an exception, additionally introduced non-canonical number types may be integrated under the `NUMBER` type in the global sort order.
+
+=== Equality ===
+
+We propose that equality and comparability must be aligned, i.e. `l = r` if and only if `l <= r && l >= r`.
+
+To achieve this, we unify the difference between equality and equivalence, by making lists containing `null` values follow the same difference as the `null` values themselves.
+
+Concretely, we propose to redefine how equality works for lists in Cypher today.
+To determine if two lists `l1` and `l2` are equal, we propose two simple tests, like so
+
+* `l1` and `l2` must have the same size, i.e. inversely `size(l1) <> size(l2>) => l1 <> l2`
+* the pairwise elements of both `l1` and `l2` must be equal, i.e.
+----
+[a1, a2, ...] = [b1, b2, ...]
+<=>
+a1 = b1 && a2 = b2 && ...
+----
+
+For clarity, we also repeat the current equality semantics of maps here (at the time of writing this CIP). Two maps `m1` and `m2` are considered equal if
+
+* `m1` and `m2` must have the same keys.
+* For each key `k`,
+** either `m1.k = m2.k` is `true`,
+** or both `m1.k IS NULL` and `m2.k IS NULL`
+
+This is at odds with the common interpretation of `null` as standing for any possible value.
+
+It is aligned though with the most common use case for maps with `null` entries which is to update multiple properties at once, e.g. `SET n += { n: 12, remove_this_key: null }`. In this case, there is no need to differentiate between different `null` values as `null` merely serves as a marker for keys to be removed. Current equality semantics makes it easy to check if two maps correspond to the same property update in this scenario.
+
+However this type of update map comparison is rare and could be emulated using a more complex predicate. The current rules do however break symmetry with how equality handles `null` in all other cases. This becomes more apparent by considering these two examples
+
+* `expr1 = expr2` => `null` if `expr1 IS null && expr2 IS NULL`
+* `{a: expr1} = {a: expr2}` => `true` (!!) if `expr1 IS null && expr2 IS NULL`
+
+To fix this, we propose that two maps `m1` and `m2` instead should be equal if
+
+* `m1` and `m2` have the same keys (the order of keys as returned by `keys` does not matter)
+* For each such key `k`, `m1.k = m2.k`
+
+As a consequence of these changes, plain equality is not reflexive in a classic sense (Consider: `{a: null} = {a: null}`, `[null] = [null]`). However this was already the case (Consider: `null = null` => `null`).
+
+However, Equality is reflexive for values that do not involve `null`.
+
+=== Equivalence ===
+
+Equivalence remains unchanged but now can be defined succinctly as being identical to equality except that:
```

Comment:
Once more, I think the next two asterisked points have no formatting.... these ought to be bullet points